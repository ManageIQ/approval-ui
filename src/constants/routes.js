const routes = {
  allrequests: {
    index: '/allrequests'
  },
  requests: {
    index: '/requests',
    addComment: '/requests/add-comment',
    approve: '/requests/approve',
    deny: '/requests/deny'
  },
  request: {
    index: '/request',
    addComment: '/request/add-comment',
    approve: '/request/approve',
    deny: '/request/deny'
  },
  workflows: {
    index: '/workflows',
    add: '/workflows/add-workflow',
    remove: '/workflows/remove',
    editInfo: '/workflows/edit-info',
    editGroups: '/workflows/edit-groups',
    editSequence: '/workflows/edit-sequence'
  }
};

export default routes;
