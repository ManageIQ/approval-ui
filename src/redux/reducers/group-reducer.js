import {
  FETCH_RBAC_GROUPS,
  FETCH_GROUP_NAMES
} from '../action-types';

// Initial State
export const groupsInitialState = {
  groups: [],
  isLoading: false
};

const setLoadingState = state => ({ ...state, isLoading: true });
const setGroups = (state, { payload }) => ({ ...state, groups: payload, isLoading: false });
const setGroupNames = (state, { payload }) => ({ ...state, groupNames: payload, isLoading: false });

export default {
  [`${FETCH_RBAC_GROUPS}_PENDING`]: setLoadingState,
  [`${FETCH_RBAC_GROUPS}_FULFILLED`]: setGroups,
  [`${FETCH_GROUP_NAMES}_PENDING`]: setLoadingState,
  [`${FETCH_GROUP_NAMES}_FULFILLED`]: setGroupNames
};
