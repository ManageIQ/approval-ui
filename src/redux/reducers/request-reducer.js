import {
  FETCH_REQUEST,
  FETCH_REQUEST_CONTENT,
  FETCH_REQUESTS,
  EXPAND_REQUEST
} from '../../redux/action-types';

// Initial State
export const requestsInitialState = {
  requests: {
    data: [],
    meta: {
      count: 0,
      limit: 10,
      offset: 0
    }
  },
  filterValue: '',
  isRequestDataLoading: false,
  expandedRequests: [],
  selectedRequest: {
    metadata: {
      user_capabilities: {}
    },
    requests: []
  }
};

const setLoadingState = state => ({ ...state, isRequestDataLoading: true, expandedRequests: []});
const setRequests = (state, { payload }) => ({ ...state, requests: payload, isRequestDataLoading: false });
const selectRequest = (state, { payload }) => ({ ...state, selectedRequest: payload, isRequestDataLoading: false });
const setRequestContent = (state, { payload }) => ({ ...state, requestContent: payload, isRequestDataLoading: false });
const setexpandRequest = (state, { payload }) => ({ ...state, expandedRequests: [ ...state.expandedRequests, payload ]});

export default {
  [`${FETCH_REQUESTS}_PENDING`]: setLoadingState,
  [`${FETCH_REQUESTS}_FULFILLED`]: setRequests,
  [`${FETCH_REQUEST}_PENDING`]: setLoadingState,
  [`${FETCH_REQUEST}_FULFILLED`]: selectRequest,
  [`${FETCH_REQUEST_CONTENT}_PENDING`]: setLoadingState,
  [`${FETCH_REQUEST_CONTENT}_FULFILLED`]: setRequestContent,
  [EXPAND_REQUEST]: setexpandRequest
};
