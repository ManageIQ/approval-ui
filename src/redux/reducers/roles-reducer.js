import { SET_USER_ROLES } from '../action-types';
import { isApprovalAdmin } from '../../helpers/shared/helpers';

export const rolesInitialState = {
  userRoles: [],
  approvalAdmin: false
};
const setUserRoles = (state, { payload }) => ({ ...state, userRoles: payload, approvalAdmin: isApprovalAdmin(payload) });

export default {
  [SET_USER_ROLES]: setUserRoles
};
