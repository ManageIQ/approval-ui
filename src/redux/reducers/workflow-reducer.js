import {
  FETCH_WORKFLOW,
  FETCH_WORKFLOWS,
  SORT_WORKFLOWS,
  SET_FILTER_WORKFLOWS,
  MOVE_SEQUENCE,
  UPDATE_WORKFLOW
} from '../../redux/action-types';

// Initial State
export const workflowsInitialState = {
  workflows: {
    data: [],
    meta: {
      count: 0,
      limit: 50,
      offset: 0
    }
  },
  workflow: {},
  filterValue: '',
  isLoading: false,
  isRecordLoading: false,
  isUpdating: 0,
  sortBy: {
    index: 2,
    property: 'sequence',
    direction: 'asc'
  }
};

const setLoadingState = state => ({ ...state, isLoading: true });
const setRecordLoadingState = state => ({ ...state, isRecordLoading: true });
const setWorkflows = (state, { payload }) => ({ ...state, workflows: payload, isLoading: false });
const selectWorkflow = (state, { payload }) => ({ ...state, workflow: payload, isRecordLoading: false });
const setSortWorkflows = (state, { payload }) => ({
  ...state,
  sortBy: payload,
  workflows: {
    ...state.workflows,
    meta: {
      ...state.workflows.meta,
      offset: 0
    }
  }
});
const setFilterValue = (state, { payload }) => ({
  ...state,
  filterValue: payload,
  workflows: {
    ...state.workflows,
    meta: {
      ...state.workflows.meta,
      offset: 0
    }
  }
});

const tranformWf = (wf, id, sequence, isMovingUp) => {
  if (wf.id === id) {
    return { ...wf, sequence };
  }

  if (wf.sequence === sequence && isMovingUp) {
    return { ...wf, sequence: wf.sequence - 1 };
  }

  if (wf.sequence === sequence && !isMovingUp) {
    return { ...wf, sequence: wf.sequence + 1 };
  }

  return wf;
};

const moveSequence = (state, { payload }) => {
  const originalProcess = state.workflows.data.find(({ id }) => id === payload.id);

  const isMovingUp = payload?.sequence > originalProcess?.sequence;

  return ({
    ...state,
    workflows: {
      ...state.workflows,
      data: state.workflows.data
      .map(wf => tranformWf(wf, payload.id, payload.sequence, isMovingUp))
      .sort((a, b) => state.sortBy.direction === 'asc' ? a.sequence - b.sequence : b.sequence - a.sequence)
    }
  });
};

const setUpdatingWorkflow = (state) => ({
  ...state,
  isUpdating: state.isUpdating + 1
});
const finishUpdatingWorkflow = (state) => ({
  ...state,
  isUpdating: state.isUpdating - 1
});

export default {
  [`${FETCH_WORKFLOWS}_PENDING`]: setLoadingState,
  [`${FETCH_WORKFLOWS}_FULFILLED`]: setWorkflows,
  [`${FETCH_WORKFLOW}_PENDING`]: setRecordLoadingState,
  [`${FETCH_WORKFLOW}_FULFILLED`]: selectWorkflow,
  [`${UPDATE_WORKFLOW}_PENDING`]: setUpdatingWorkflow,
  [`${UPDATE_WORKFLOW}_FULFILLED`]: finishUpdatingWorkflow,
  [`${UPDATE_WORKFLOW}_REJECTED`]: finishUpdatingWorkflow,
  [SORT_WORKFLOWS]: setSortWorkflows,
  [SET_FILTER_WORKFLOWS]: setFilterValue,
  [MOVE_SEQUENCE]: moveSequence
};
