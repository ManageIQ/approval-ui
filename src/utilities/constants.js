import {
  CheckCircleIcon,
  InfoCircleIcon,
  ErrorCircleOIcon,
  ExclamationCircleIcon
} from '@patternfly/react-icons';
import React from 'react';
import requestsMessages from '../messages/requests.messages';

export const APPROVAL_API_BASE = `${process.env.BASE_PATH}/approval/v1.2`;
export const RBAC_API_BASE = `${process.env.BASE_PATH}/rbac/v1`;

export const decisionValues = {
  undecided: { displayName: requestsMessages.needsReview, color: 'blue', icon: <InfoCircleIcon /> },
  approved: { displayName: requestsMessages.approved, color: 'green', icon: <CheckCircleIcon /> },
  denied: { displayName: requestsMessages.denied, color: 'red', icon: <ExclamationCircleIcon /> },
  canceled: { displayName: requestsMessages.canceled, color: 'red', icon: <ErrorCircleOIcon /> },
  error: { displayName: requestsMessages.error, color: 'red', icon: <ExclamationCircleIcon /> }
};

// React intl does not support empty strings
export const untranslatedMessage = (defaultMessage = ' ') => ({ id: 'untranslated', defaultMessage });
export const APP_DISPLAY_NAME = {
  catalog: 'Automation Services Catalog',
  topology: 'Topological inventory'
};
