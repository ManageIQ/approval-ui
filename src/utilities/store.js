
import promiseMiddleware from 'redux-promise-middleware';
import ReducerRegistry, { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/files/ReducerRegistry';
import { notifications, notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';

import reduxLogger from 'redux-logger';
import thunk from 'redux-thunk';
import requestReducer, { requestsInitialState } from '../redux/reducers/request-reducer';
import workflowReducer, { workflowsInitialState } from '../redux/reducers/workflow-reducer';
import groupReducer, { groupsInitialState } from '../redux/reducers/group-reducer';
import rolesReducer, { rolesInitialState } from '../redux/reducers/roles-reducer';

const registry = new ReducerRegistry({}, [ thunk, promiseMiddleware(), notificationsMiddleware({
  errorTitleKey: [ 'message' ],
  errorDescriptionKey: [ 'errors', 'stack' ]
}), reduxLogger ]);

registry.register({
  requestReducer: applyReducerHash(requestReducer, requestsInitialState),
  workflowReducer: applyReducerHash(workflowReducer, workflowsInitialState),
  groupReducer: applyReducerHash(groupReducer, groupsInitialState),
  rolesReducer: applyReducerHash(rolesReducer, rolesInitialState),
  notifications
});

export default registry.getStore();
