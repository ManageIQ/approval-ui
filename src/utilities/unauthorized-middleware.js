import { approvalHistory } from '../router';

const unAuthorizedMiddleware = () => (dispatch) => (action) => {
  const nextAction = { ...action };
  if (action.type.match(/_REJECTED$/) && action?.payload?.redirect) {
    setTimeout(() => {
      approvalHistory.push(action.payload.redirect.pathname, {
        from: approvalHistory.location
      });
    });
    return;
  }

  return dispatch(nextAction);
};

export default unAuthorizedMiddleware;
