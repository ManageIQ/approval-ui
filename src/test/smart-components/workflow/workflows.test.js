import React from 'react';
import { act } from 'react-dom/test-utils';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { MemoryRouter, Route } from 'react-router-dom';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store' ;
import { shallowToJson } from 'enzyme-to-json';
import promiseMiddleware from 'redux-promise-middleware';
import { IntlProvider } from 'react-intl';

import Workflows from '../../../smart-components/workflow/workflows';
import { workflowsInitialState } from '../../../redux/reducers/workflow-reducer';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications';
import { groupsInitialState } from '../../../redux/reducers/group-reducer';
import { APPROVAL_API_BASE, RBAC_API_BASE } from '../../../utilities/constants';
import EditWorkflowInfoModal from '../../../smart-components/workflow/edit-workflow-info-modal';
import EditWorkflowStagesModal from '../../../smart-components/workflow/edit-workflow-stages-modal';
import RemoveWorkflowModal from '../../../smart-components/workflow/remove-workflow-modal';
import AddStagesWizard from '../../../smart-components/workflow/add-stages/add-stages-wizard';

const ComponentWrapper = ({ store, initialEntries = [ '/workflows' ], children }) => (
  <Provider store={ store }>
    <MemoryRouter initialEntries={ initialEntries }>
      <IntlProvider locale="en">
        { children }
      </IntlProvider>
    </MemoryRouter>
  </Provider>
);

describe('<Workflows />', () => {

  let initialProps;
  const middlewares = [ thunk, promiseMiddleware(), notificationsMiddleware() ];
  let mockStore;
  let initialState;
  let stateWithData;

  beforeEach(() => {
    initialProps = {
      history: {
        goBack: jest.fn(),
        push: jest.fn()
      }
    };
    mockStore = configureStore(middlewares);
    initialState = { workflowReducer: { ...workflowsInitialState, isLoading: false }, groupReducer: { ...groupsInitialState }};
    stateWithData = {
      groupReducer: { ...groupsInitialState },
      workflowReducer: {
        ...workflowsInitialState,
        data: [{
          id: '123',
          name: 'foo'
        }]
      }
    };
  });

  it('should render correctly', () => {
    const store = mockStore(initialState);
    const wrapper = shallow(<Workflows store={ store } { ...initialProps } />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should render correctly in loading state', () => {
    const store = mockStore(initialState);
    const wrapper = shallow(<Workflows store={ store } { ...initialProps } isLoading />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should redirect to Edit info page', async done => {
    const store = mockStore(stateWithData);
    let wrapper;

    apiClientMock.get(`${APPROVAL_API_BASE}/workflows?limit=10&offset=0`, mockOnce({ body: { data: [{
      id: 'edit-id',
      name: 'foo',
      group_refs: [ 'group-1' ],
      group_names: [ 'group-name-1' ]
    }]}}));
    apiClientMock.get(`${RBAC_API_BASE}/groups/`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${RBAC_API_BASE}/groups/group-1/`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${APPROVAL_API_BASE}/workflows/edit-id`, mockOnce({ body: { data: []}}));
    await act(async()=> {
      wrapper = mount(
        <ComponentWrapper store={ store }>
          <Route path="/workflows" component={ Workflows } />
        </ComponentWrapper>
      );
    });
    wrapper.update();
    /**
     * Open action drop down and click on edit action
     */
    wrapper.find('button.pf-c-dropdown__toggle.pf-m-plain').last().simulate('click');
    await act(async() => {
      wrapper.find('a.pf-c-dropdown__menu-item').first().simulate('click');
    });

    wrapper.update();
    expect(wrapper.find(MemoryRouter).instance().history.location.pathname).toEqual('/workflows/edit-info/edit-id');
    expect(wrapper.find(EditWorkflowInfoModal)).toHaveLength(1);
    done();
  });

  it('should redirect to Edit workflow stages page', async done => {
    const store = mockStore(stateWithData);
    let wrapper;

    apiClientMock.get(`${APPROVAL_API_BASE}/workflows?limit=10&offset=0`, mockOnce({ body: { data: [{
      id: 'edit-id',
      name: 'foo',
      group_refs: [ 'group-1' ],
      group_names: [ 'group-name-1' ]
    }]}}));
    apiClientMock.get(`${RBAC_API_BASE}/groups/`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${RBAC_API_BASE}/groups/group-1/`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${APPROVAL_API_BASE}/workflows/edit-id`, mockOnce({ body: { group_refs: []}}));
    await act(async()=> {
      wrapper = mount(
        <ComponentWrapper store={ store }>
          <Route path="/workflows" component={ Workflows } />
        </ComponentWrapper>
      );
    });
    wrapper.update();
    /**
     * Open action drop down and click on edit stages action
     */
    wrapper.find('button.pf-c-dropdown__toggle.pf-m-plain').last().simulate('click');
    await act(async() => {
      wrapper.find('a.pf-c-dropdown__menu-item').at(1).simulate('click');
    });

    wrapper.update();
    expect(wrapper.find(MemoryRouter).instance().history.location.pathname).toEqual('/workflows/edit-stages/edit-id');
    expect(wrapper.find(EditWorkflowStagesModal)).toHaveLength(1);
    done();
  });

  it('should redirect to Delete workflow page', async done => {
    const store = mockStore(stateWithData);
    let wrapper;

    apiClientMock.get(`${APPROVAL_API_BASE}/workflows?limit=10&offset=0`, mockOnce({ body: { data: [{
      id: 'edit-id',
      name: 'foo',
      group_refs: [ 'group-1' ],
      group_names: [ 'group-name-1' ]
    }]}}));
    apiClientMock.get(`${RBAC_API_BASE}/groups/`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${RBAC_API_BASE}/groups/group-1/`, mockOnce({ body: { data: []}}));
    await act(async()=> {
      wrapper = mount(
        <ComponentWrapper store={ store }>
          <Route path="/workflows" component={ Workflows } />
        </ComponentWrapper>
      );
    });
    wrapper.update();
    /**
     * Open action drop down and click on delete action
     */
    wrapper.find('button.pf-c-dropdown__toggle.pf-m-plain').last().simulate('click');
    await act(async() => {
      wrapper.find('a.pf-c-dropdown__menu-item').at(2).simulate('click');
    });

    wrapper.update();
    expect(wrapper.find(MemoryRouter).instance().history.location.pathname).toEqual('/workflows/remove/edit-id');
    expect(wrapper.find(RemoveWorkflowModal)).toHaveLength(1);
    done();
  });

  it('should redirect to add workflow page', async done => {
    const store = mockStore(stateWithData);
    let wrapper;

    apiClientMock.get(`${APPROVAL_API_BASE}/workflows?limit=10&offset=0`, mockOnce({ body: { data: [{
      id: 'edit-id',
      name: 'foo',
      group_refs: [ 'group-1' ],
      group_names: [ 'group-name-1' ]
    }]}}));
    apiClientMock.get(`${RBAC_API_BASE}/groups/`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${RBAC_API_BASE}/groups/group-1/`, mockOnce({ body: { data: []}}));
    await act(async()=> {
      wrapper = mount(
        <ComponentWrapper store={ store }>
          <Route path="/workflows" component={ Workflows } />
        </ComponentWrapper>
      );
    });
    wrapper.update();
    /**
     * Click on add workflow link
     */
    wrapper.find('Link#add-workflow-link').simulate('click', { button: 0 });

    wrapper.update();
    expect(wrapper.find(MemoryRouter).instance().history.location.pathname).toEqual('/workflows/add-workflow');
    expect(wrapper.find(AddStagesWizard)).toHaveLength(1);
    done();
  });

  it('should remove multiple selected workflows from table', async done => {
    const store = mockStore(stateWithData);
    let wrapper;

    apiClientMock.get(`${APPROVAL_API_BASE}/workflows?limit=10&offset=0`, mockOnce({ body: { data: [{
      id: 'edit-id',
      name: 'foo',
      group_refs: [ 'group-1' ],
      group_names: [ 'group-name-1' ]
    }]}}));
    apiClientMock.get(`${RBAC_API_BASE}/groups/`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${RBAC_API_BASE}/groups/group-1/`, mockOnce({ body: { data: []}}));
    await act(async()=> {
      wrapper = mount(
        <ComponentWrapper store={ store }>
          <Route path="/workflows" component={ Workflows } />
        </ComponentWrapper>
      );
    });
    wrapper.update();
    wrapper.find('input[type="checkbox"]').last().simulate('change', { target: { checked: true }});
    wrapper.find('Link#remove-multiple-workflows').simulate('click', { button: 0 });
    wrapper.update();
    expect(wrapper.find(MemoryRouter).instance().history.location.pathname).toEqual('/workflows/remove');
    expect(wrapper.find(RemoveWorkflowModal)).toHaveLength(1);
    done();
  });
});
