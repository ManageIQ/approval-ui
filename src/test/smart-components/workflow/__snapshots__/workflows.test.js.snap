// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Workflows /> should render correctly 1`] = `
<ContextProvider
  value={null}
>
  <Workflows
    fetchRbacGroups={[Function]}
    fetchWorkflows={[Function]}
    history={
      Object {
        "goBack": [MockFunction],
        "push": [MockFunction],
      }
    }
    isLoading={false}
    pagination={
      Object {
        "count": 0,
        "limit": 10,
        "offset": 0,
      }
    }
    rbacGroups={Array []}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    workflows={Array []}
  />
</ContextProvider>
`;

exports[`<Workflows /> should render correctly in loading state 1`] = `
<ContextProvider
  value={null}
>
  <Workflows
    fetchRbacGroups={[Function]}
    fetchWorkflows={[Function]}
    history={
      Object {
        "goBack": [MockFunction],
        "push": [MockFunction],
      }
    }
    isLoading={false}
    pagination={
      Object {
        "count": 0,
        "limit": 10,
        "offset": 0,
      }
    }
    rbacGroups={Array []}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    workflows={Array []}
  />
</ContextProvider>
`;
